compiler/src/dotty/tools/dotc/core/classfile/ClassfileParser.scala

```diff
@@ -237,70 +241,66 @@ class ClassfileParser(
 
   val memberCompleter: LazyType = new LazyType {
 
-    def complete(denot: SymDenotation)(using Context): Unit = {
-      val oldbp = in.bp
-      try {
-        in.bp = denot.symbol.coord.toIndex
-        val sym = denot.symbol
-        val jflags = in.nextChar
-        val isEnum = (jflags & JAVA_ACC_ENUM) != 0
-        val name = pool.getName(in.nextChar)
-        val isConstructor = name eq nme.CONSTRUCTOR
-
-        /** Strip leading outer param from constructor and trailing access tag for
-         *  private inner constructors.
-         */
-        def normalizeConstructorParams() = innerClasses.get(currentClassName) match {
-          case Some(entry) if !isStatic(entry.jflags) =>
-            val mt @ MethodTpe(paramNames, paramTypes, resultType) = denot.info
-            var normalizedParamNames = paramNames.tail
-            var normalizedParamTypes = paramTypes.tail
-            if ((jflags & JAVA_ACC_SYNTHETIC) != 0) {
-              // SI-7455 strip trailing dummy argument ("access constructor tag") from synthetic constructors which
-              // are added when an inner class needs to access a private constructor.
-              normalizedParamNames = paramNames.dropRight(1)
-              normalizedParamTypes = paramTypes.dropRight(1)
-            }
-            denot.info = mt.derivedLambdaType(normalizedParamNames, normalizedParamTypes, resultType)
-          case _ =>
-        }
-
-        /** Make return type of constructor be the enclosing class type,
-         *  and make constructor type polymorphic in the type parameters of the class
-         */
-        def normalizeConstructorInfo() = {
-          val rt = classRoot.typeRef appliedTo (classRoot.typeParams map (_.typeRef))
-
-          def resultType(tpe: Type): Type = tpe match {
-            case mt @ MethodType(paramNames) => mt.derivedLambdaType(paramNames, mt.paramInfos, rt)
-            case pt : PolyType => pt.derivedLambdaType(pt.paramNames, pt.paramInfos, resultType(pt.resType))
+    def complete(denot: SymDenotation)(using Context): Unit = withReader(classfile) { (using in) =>
+      in.bp = denot.symbol.coord.toIndex
+      val sym = denot.symbol
+      val jflags = in.nextChar
+      val isEnum = (jflags & JAVA_ACC_ENUM) != 0
+      val name = pool.getName(in.nextChar).name
```

Comment:
This does not seem to match what scalac did in https://github.com/scala/scala/pull/6653/files where `DataReader#using` is only called at the toplevel, here we're also calling it every time we complete a symbol which sounds expensive. To avoid that in Scala 2 a MemberTypeComplter takes as input the name, jflags and descriptor so it doesn't need to read the file again.

Reply:
In Scala 2, it eagerly reads all attributes of members, here we delay reading attributes. 
From the statistics I collected for compiling `typer/*.scala`, only 1/10 members are completed.
It means we might call `parseAttributes` ~2400 times unnecessarily.

-  member.complete: 283
-  parseMember: 2725
- parseClass: 652

Another reason I'm a little hesitant to follow exactly what is done in Scala 2 is that the code becomes more complex, while it only benefits Java methods.

On the other hand, indeed there is a concern here: delaying read involves additional IO on the file, which could possibly cancel the algorithmic savings (disk buffering might alleviate a little the problem). Re-opening + read class files 283 times is not ideal, we definitely need an improvement here.

I'll play more with it and see what is the best we can do here.

Reply:
> delaying read involves additional IO on the file, which could possibly cancel the algorithmic savings (disk buffering might alleviate a little the problem).

Unfortunately we can never rely on IO being fast enough, it is too dependent on the system, opening a file is especially problematic since it involves at least one system call which are generally slow, and it's a disaster on Windows because apparently every time you open a file the antivirus and other system services is going to run some code on the file to see what it does. So I really think we should follow what Scala 2 did here (or keep what we currently have, but that's worse because it involves keeping the content of files in memory as long as we might need to complete something from them).

Reply:
My main worry here is that the code gets less maintainable, while the benefits are unclear (due to the 10x more work in reading attributes than necessary).

A better alternative might be to just load members eagerly, as the member completer does very few things, compared to attribute reading. This way, we avoid duplicate IO and make the code maintainable.

Reply:
> due to the 10x more work in reading attributes than necessary)

Since the file is already loaded in memory and we only need to extract a few strings and integers for every member, I don't expect that work to be too expensive.

Reply:
Now we parse attributes eagerly while minimizing the changes to the code.

Reply:
> Now we parse attributes eagerly while minimizing the changes to the code.

Having to create and retrieve all these closures looks both complicated and expensive to me. Why is this needed here but not in scalac?

Reply:
The whole purpose is to port the benefits while without losing maintainability of the code. If there is evidence that this is a performance issue, we can easily specialize the logic in `AttributeCompileter` as it's done for `exceptions`.

Reply:
I agree that maintainability is important but I don't understand the trade-off here, what is less maintainable about the way that scalac does things here?

Reply:
In our implementation, we completely avoided creating the hierarchy:

```Scala
abstract class JavaTypeCompleter extends LazyType
class ClassTypeCompleter(name: Name, jflags: JavaAccFlags, parent: NameOrString, ifaces: List[NameOrString]) extends JavaTypeCompleter
class MemberTypeCompleter(name: Name, jflags: JavaAccFlags, descriptor: String) extends JavaTypeCompleter
```

Unlike Scala2, we don't really need `ClassTypeCompleter` here. Because in Dotty the `ClassfileParser` is already called from a `LazyType`. It does not make sense to return another `LazyType` in lazy type completion.

Therefore, we need to minimize the changes to `parseAttributes` so that it can be used both for Class attribute parsing and member attributes parsing (which has to be partially lazy).

We introduced `AttributeCompleter` to minimize the changes to our existing `parseAttributes` and `memberCompleter`.

If closure creation is a concern, we can simply specialize more things directly in `AttributeCompleter`. It can be easily handled in the current design.

Reply:
Thanks for the explanation.

> Therefore, we need to minimize the changes to parseAttributes so that it can be used both for Class attribute parsing and member attributes parsing (which has to be partially lazy).

OK, so it seems that in scalac, parseAttributes is always eager but in dotty currently it's called eagerly at the class-level but lazily for members.

> Therefore, we need to minimize the changes to parseAttributes so that it can be used both for Class attribute parsing and member attributes parsing (which has to be partially lazy).

I see, that makes sense, but I think this can be achieved without making parseAttributes partially lazy. Basically, the input to parseAttributes is an Array of bytes (wrapped in a DataReader), we can keep calling parseAttributes in the member completer if we pass as input to the completer an array of bytes containing the attributes (so the only thing we do eagerly is copy these bytes into an array), WDYT?

Reply:
> Basically, the input to parseAttributes is an Array of bytes (wrapped in a DataReader), we can keep calling parseAttributes in the member completer if we pass as input to the completer an array of bytes containing the attributes (so the only thing we do eagerly is copy these bytes into an array), WDYT?

That is another design idea that I tried. It creates more complexity because the attributes may contain references to constant pools, where the constants are loaded lazily. Therefore, I gave up pursuing the design.

BTW, in 940b2f1, I've removed all closures.

Reply:
> It creates more complexity because the attributes may contain references to constant pools, where the constants are loaded lazily.

This makes sense, this should be documented somewhere since it's easy to miss, I suggest explaining that in a documentation comment for AttributeCompleter.

Reply:
Added in 1ddb130.